scrapy_poet/utils.py

```diff
@@ -102,14 +102,12 @@ def discover_callback_providers(callback):
     plan = andi.plan(
         callback,
         is_injectable=is_injectable,
-        externally_provided=providers.keys(),
+        externally_provided=set.union(*(p.provided_classes for p in providers)),
     )
     for obj, _ in plan:
-        provider = providers.get(obj)
-        if not provider:
-            continue
-
-        yield provider
+        for provider in providers:
```

Comment:
With providers being a set an order is a bit arbitrary now. Does it make sense to change it to a list, to ensure priorities are explicit?

Reply:
- sets are a good option because they enforce uniqueness 
- lists on the other hand could make it possible to play with providers' order but
  - they don't guarantee uniqueness, what could lead to problems (maybe)
  - we don't depend on providers' order at the moment, so we can probably postpone this task

Reply:
Not sure about your example. 

Isn't it better to override the original AutoExtract provider and register / install this variation instead? I suppose it would lead to less bugs and would give better accountability for which providers are responsible for what.

Reply:
@victor-torres there is an example above https://github.com/scrapinghub/scrapy-poet/pull/40#discussion_r492556338 . 

An alternative would make providers disjoint. That is, you don't allow having two providers providing the same class, and if this happens you raise an error. I think this is what you are thinking on. And then ordering is not important because is impossible for two providers to provide objects of the same class. 

So let's go back to an example I put before. In the case you have a provider like:
```py
class AutoExtractProvider(PageObjectInputProvider):
  provided_classes = [Article, Product, JobPosting, Html]

  def __init__(request: DummyRequest, stats: StatsCollector)

  def __call__(...):
    ....
```

If you want to override let's say product, you would have to write your own provider:

```py
class OverridenAutoExtractProvider(PageObjectInputProvider):
  provided_classes = AutoExtractProvider.provided_classes

  # Note that you have to mimic the annotations from the original provider so that the injection works. 
  def __init__(request: DummyRequest, stats: StatsCollector):
    self.origina_provider = AutoExtractProvider(request, stats)
    ...

  def __call__(to_provide)
    # here you have to provide product using your own custom code, and then the rest calling to
    # self.original_provider.__call__, put all together, and return it. 
  ....
```
And then replace the old one with the new one. 

I see now better the idea :+1: . Still not sure about it. I don't see the difficulties of having a ordered list of providers and keep the `to_provide_remaining` list. And it would make the overriding simpler.